File: app/src/main/java/com/midburn/gate/midburngate/activities/ShowActivity.java
package com.midburn.gate.midburngate.activities;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.NavUtils;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.midburn.gate.midburngate.HttpRequestListener;
import com.midburn.gate.midburngate.R;
import com.midburn.gate.midburngate.application.MainApplication;
import com.midburn.gate.midburngate.consts.AppConsts;
import com.midburn.gate.midburngate.model.Group;
import com.midburn.gate.midburngate.model.Ticket;
import com.midburn.gate.midburngate.utils.AppUtils;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;

import okhttp3.HttpUrl;
import okhttp3.Response;

public class ShowActivity
		extends AppCompatActivity {

	private static final int PRODUCTION_EARLY_ARRIVAL_WORKAROUND = -1;
	private TextView     mTicketOrderNumberTextView;
	private TextView     mTicketNumberTextView;
	private TextView     mTicketOwnerNameTextView;
	private TextView     mTicketTypeTextView;
	private LinearLayout mDisabledLayout;
	private TextView     mTicketOwnerIdTextView;

	private Button      mEntranceButton;
	private Button      mExitButton;
	private Button      mCancelButton;
	private ProgressBar mProgressBar;

	private enum State {
		ERALY_ENTRANCE,
		MIDBURN
	}

	private enum Action {
		ENTER,
		EXIT
	}

	private State mState;
	private Action mAction;

	private HttpRequestListener mHttpRequestListener;

	private String mGateCode;
	private Ticket mTicket;
	private Group mSelectedGroup;

	public void exit(View view) {
		boolean hasInternetConnection = AppUtils.isConnected(this);
		if (!hasInternetConnection) {
			AppUtils.createAndShowDialog(this, getString(R.string.no_network_dialog_title), getString(R.string.no_network_dialog_message), getString(R.string.ok), null, null, null, android.R.drawable.ic_dialog_alert);
			return;
		}
		if (mTicket == null) {
			Log.e(AppConsts.TAG, "ticket is null");
			return;
		}

		mProgressBar.setVisibility(View.VISIBLE);

		String barcode = mTicket.getBarCode();
		Log.d(AppConsts.TAG, "user barcode to exit: " + barcode);

		HttpUrl url = new HttpUrl.Builder().scheme("https")
		                                   .host(AppConsts.SERVER_URL)
		                                   .addPathSegment("api")
		                                   .addPathSegment("gate")
		                                   .addPathSegment("gate-exit")
		                                   .build();

		JSONObject jsonObject = new JSONObject();
		try {
			jsonObject.put("event_id", mGateCode);
			jsonObject.put("barcode", barcode);
		} catch (JSONException e) {
			Log.e(AppConsts.TAG, e.getMessage());
		}

		AppUtils.doPOSTHttpRequest(url, jsonObject.toString(), mHttpRequestListener);
	}

	public void entrance(View view) {
		boolean hasInternetConnection = AppUtils.isConnected(this);
		if (!hasInternetConnection) {
			AppUtils.createAndShowDialog(this, getString(R.string.no_network_dialog_title), getString(R.string.no_network_dialog_message), getString(R.string.ok), null, null, null, android.R.drawable.ic_dialog_alert);
			return;
		}
		if (mTicket == null) {
			Log.e(AppConsts.TAG, "ticket is null");
			return;
		}
		if (mState.equals(State.ERALY_ENTRANCE)) {
			handleGroupTypes();
		}
		else if (mState.equals(State.MIDBURN)) {
			sendEntranceRequestWithoutGroups();
		}
		else {
			Log.e(AppConsts.TAG, "unknown state. mState: " + mState);
		}
	}

	private void handleGroupTypes() {
		final ArrayList<Group> groupsArrayList = mTicket.getGroups();

		//check if group type is production. if so, select it immediately
		if (mTicket.isProductionEarlyArrival) {
			sendEntranceRequest(PRODUCTION_EARLY_ARRIVAL_WORKAROUND);
			return;
		}

		// no groups alert
		if (groupsArrayList == null || groupsArrayList.size() == 0) {
			AppUtils.createAndShowDialog(this, "שגיאה", getString(R.string.no_early_arrival_message), getString(R.string.ok), null, null, null, android.R.drawable.ic_dialog_alert);
			return;
		}

		// show group selection dialog
		int groupsArrayListSize = groupsArrayList.size();
		CharSequence groupsArray[] = new CharSequence[groupsArrayListSize];
		for (int i = 0 ; i < groupsArrayListSize ; i++) {
			Group group = groupsArrayList.get(i);
			groupsArray[i] = getGroupType(group) + ": "+ group.getName();
		}
		AlertDialog.Builder builder = new AlertDialog.Builder(this);
		builder.setTitle("בחר קבוצה");
		builder.setItems(groupsArray, new DialogInterface.OnClickListener() {
			@Override
			public void onClick(DialogInterface dialog, int which) {
				Group selectedGroup = groupsArrayList.get(which);
				Log.d(AppConsts.TAG, selectedGroup.getName() + " was clicked. id: " + selectedGroup.getId());
				mProgressBar.setVisibility(View.VISIBLE);
				mSelectedGroup = selectedGroup;
				sendEntranceRequest(selectedGroup.getId());
			}
		});
		builder.show();
	}

	private String getGroupType(Group group) {
		if (group.getType() == null) {
			return "";
		}

		if (group.getType().equals(AppConsts.GROUP_TYPE_ART)) {
			return "מיצב";
		}

		if (group.getType().equals(AppConsts.GROUP_TYPE_CAMP)) {
			return "מחנה";
		}

		if (group.getType().equals(AppConsts.GROUP_TYPE_PRODUCTION)) {
			return "הפקה";
		}

		return "";
	}

	private void sendEntranceRequestWithoutGroups() {
		String barcode = mTicket.getBarCode();
		Log.d(AppConsts.TAG, "user barcode to enter: " + barcode);
		JSONObject jsonObject = new JSONObject();
		try {
			jsonObject.put("barcode", barcode);
			jsonObject.put("event_id", mGateCode);
		} catch (JSONException e) {
			Log.e(AppConsts.TAG, e.getMessage());
		}
		HttpUrl url = new HttpUrl.Builder().scheme("https")
		                                   .host(AppConsts.SERVER_URL)
		                                   .addPathSegment("api")
		                                   .addPathSegment("gate")
		                                   .addPathSegment("gate-enter")
		                                   .build();

		AppUtils.doPOSTHttpRequest(url, jsonObject.toString(), mHttpRequestListener);
	}

	private void sendEntranceRequest(int groupId) {
		String barcode = mTicket.getBarCode();
		Log.d(AppConsts.TAG, "user barcode to enter: " + barcode);
		JSONObject jsonObject = new JSONObject();
		try {
			jsonObject.put("barcode", barcode);
			jsonObject.put("event_id", mGateCode);
			// TODO(alex): 5/7/18 this is a workaround -1 means this is a production early arrival. server will validate that if no group is present
			if (groupId != PRODUCTION_EARLY_ARRIVAL_WORKAROUND) {
				jsonObject.put("group_id", groupId);
			}
		} catch (JSONException e) {
			Log.e(AppConsts.TAG, e.getMessage());
		}
		HttpUrl url = new HttpUrl.Builder().scheme("https")
		                                   .host(AppConsts.SERVER_URL)
		                                   .addPathSegment("api")
		                                   .addPathSegment("gate")
		                                   .addPathSegment("gate-enter")
		                                   .build();

		AppUtils.doPOSTHttpRequest(url, jsonObject.toString(), mHttpRequestListener);
	}

	public void cancel(View view) {
		Intent intent = new Intent(ShowActivity.this, MainActivity.class);
		startActivity(intent);
	}

	private void handleServerResponse(final Response response) {
		MainApplication.getsMainThreadHandler()
		               .post(() -> {
			               if (response == null) {
				               Log.e(AppConsts.TAG, "response is null");
				               AppUtils.playMusic(ShowActivity.this, AppConsts.ERROR_MUSIC);
				               AppUtils.createAndShowDialog(ShowActivity.this, "פעולה נכשלה", null, getString(R.string.ok), null, null, null, android.R.drawable.ic_dialog_alert);
				               return;
			               }
			               try {
				               String responseBodyString = response.body()
				                                                   .string();
				               Log.d(AppConsts.TAG, "response.body():" + responseBodyString);
				               if (response.code() == AppConsts.RESPONSE_OK) {
					               JSONObject jsonObject = new JSONObject(responseBodyString);
					               AppUtils.playMusic(ShowActivity.this, AppConsts.OK_MUSIC);
					               String resultMessage = (String) jsonObject.get("message");
					               Log.d(AppConsts.TAG, "resultMessage: " + resultMessage);
								   showConfirmationAlert();
				               }
				               else {
					               Log.e(AppConsts.TAG, "response code: " + response.code() + " | response body: " + responseBodyString);
					               AppUtils.playMusic(ShowActivity.this, AppConsts.ERROR_MUSIC);
					               JSONObject jsonObject = new JSONObject(responseBodyString);
					               String errorMessage = (String) jsonObject.get("error");
					               AppUtils.createAndShowDialog(ShowActivity.this, "שגיאה", AppUtils.getErrorMessage(ShowActivity.this, errorMessage), getString(R.string.ok), null, null, null, android.R.drawable.ic_dialog_alert);
				               }
			               } catch (IOException | JSONException e) {
				               Log.e(AppConsts.TAG, e.getMessage());
				               AppUtils.playMusic(ShowActivity.this, AppConsts.ERROR_MUSIC);
				               AppUtils.createAndShowDialog(ShowActivity.this, "שגיאה", e.getMessage(), getString(R.string.ok), null, null, null, android.R.drawable.ic_dialog_alert);
			               }
		               });
	}

	private  void showConfirmationAlert() {
		String message = "";

		if (mAction == Action.ENTER) {
			message = mTicket.getTicketOwnerName() + " נכנס/ה בהצלחה לאירוע.";
			if (mState == State.ERALY_ENTRANCE) {
				message += "\n" + "הקצאה לכניסה מוקדמת - " + (mSelectedGroup != null ? mSelectedGroup
						.getName() : "הפקה");
			}
		} else {
			message = mTicket.getTicketOwnerName() + " יצא/ה בהצלחה מהאירוע.";
		}

		AlertDialog.Builder builder = new AlertDialog.Builder(ShowActivity.this);
		builder.setMessage(message)
				.setTitle("אישור")
				.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialog, int which) {
						Intent intent = new Intent(ShowActivity.this, MainActivity.class);
						startActivity(intent);
					}
				});
		AlertDialog dialog = builder.create();
		dialog.show();
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_show);
		getSupportActionBar().setTitle(getString(R.string.ticket_details));
		bindView();

		mGateCode = getIntent().getStringExtra("event_id");
		if (TextUtils.isEmpty(mGateCode)) {
			mGateCode = AppUtils.getEventId(this);
		}
		mHttpRequestListener = new HttpRequestListener() {
			@Override
			public void onResponse(Response response) {
				runOnUiThread(new Runnable() {
					@Override
					public void run() {
						mProgressBar.setVisibility(View.GONE);
					}
				});
				handleServerResponse(response);
			}
		};

		Ticket ticket = (Ticket) getIntent().getSerializableExtra("ticketDetails");
		if (ticket != null) {

			mTicket = ticket;
			mTicketOrderNumberTextView.setText(String.valueOf(ticket.getInvitationNumber()));
			mTicketNumberTextView.setText(String.valueOf(ticket.getTicketNumber()));
			mTicketOwnerNameTextView.setText(ticket.getTicketOwnerName());
			mTicketTypeTextView.setText(ticket.getTicketType());
			mTicketOwnerIdTextView.setText(ticket.getTicketOwnerId());

			if (mTicket.isProductionEarlyArrival) {
				findViewById(R.id.earlyArrivalProductionTV).setVisibility(View.VISIBLE);
			}

			//decide which button to show (entrance/exit)
			if (ticket.isInsideEvent() == 0) {
				//the user is outside the event
				mAction = Action.ENTER;
			}
			else if (ticket.isInsideEvent() == 1) {
				//the user is inside the event
				mAction = Action.EXIT;
			}
			else {
				Log.e(AppConsts.TAG, "unknown isInsideEvent state. isInsideEvent: " + ticket.isInsideEvent());
			}
			toggleButtonsState();


			//decide if disabled layout should be displayed
			if (ticket.getIsDisabled() == 1) {
				//show disabled parking
				mDisabledLayout.setVisibility(View.VISIBLE);
			}
			else {
				mDisabledLayout.setVisibility(View.GONE);
			}

			//early arrival mode
			if (ticket.getGateStatus() != null && ticket.getGateStatus().equals("early_arrival")) {
				mState = State.ERALY_ENTRANCE;
			}else { //otherwise, this is the real deal
				mState = State.MIDBURN;
			}
		}
	}

	private void toggleButtonsState() {
		if (mAction == Action.EXIT) {
			mEntranceButton.setVisibility(View.GONE);
			mExitButton.setVisibility(View.VISIBLE);
		}
		else {
			mExitButton.setVisibility(View.GONE);
			mEntranceButton.setVisibility(View.VISIBLE);
		}
	}

	private void bindView() {
		mTicketOrderNumberTextView = findViewById(R.id.orderNumberTextView_ShowActivity);
		mTicketNumberTextView = findViewById(R.id.ticketNumberTextView_ShowActivity);
		mTicketOwnerNameTextView = findViewById(R.id.ticketOwnerTextView_ShowActivity);
		mTicketTypeTextView = findViewById(R.id.ticketTypeTextView_ShowActivity);
		mEntranceButton = findViewById(R.id.entranceButton_ShowActivity);
		mExitButton = findViewById(R.id.exitButton_ShowActivity);
		mCancelButton = findViewById(R.id.cancelButton_ShowActivity);
		mProgressBar = findViewById(R.id.progressBar_ShowActivity);
		mTicketOwnerIdTextView = findViewById(R.id.ticketOwnerIdTextView_ShowActivity);
		mDisabledLayout = findViewById(R.id.disabledLayout_ShowActivity);
	}

	@Override
	public void onBackPressed() {
		Intent upIntent = new Intent(this, MainActivity.class);
		NavUtils.navigateUpTo(this, upIntent);
	}
}

File: app/src/main/java/com/midburn/gate/midburngate/activities/ShowActivity.kt
package com.midburn.gate.midburngate.activities

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.NavUtils
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import com.midburn.gate.midburngate.R
import com.midburn.gate.midburngate.consts.AppConsts
import com.midburn.gate.midburngate.network.NetworkApi
import com.midburn.gate.midburngate.network.TicketNew
import com.midburn.gate.midburngate.utils.AppUtils
import kotlinx.android.synthetic.main.activity_show.*

class ShowActivity : AppCompatActivity() {
    private var mAction: Action? = null

    private lateinit var mGateCode: String
    private lateinit var mTicket: TicketNew
    private var mSelectedGroup: com.midburn.gate.midburngate.network.Group? = null

    private val gateCallback = object : NetworkApi.Callback<Unit> {
        override fun onSuccess(response: Unit) {
            runOnUiThread { progressBar_ShowActivity.visibility = View.GONE }

            AppUtils.playMusic(this@ShowActivity, AppConsts.OK_MUSIC)
            var message: String
            if (mAction == Action.ENTER) {
                message = mTicket.ticket.holder_name + " נכנס/ה בהצלחה לאירוע."
                if (mTicket.gate_status == TicketNew.State.EARLY_ENTRANCE) {
                    message += "\n" + "הקצאה לכניסה מוקדמת - " + if (mSelectedGroup != null)
                        mSelectedGroup!!
                                .name
                    else
                        "הפקה"
                }
            } else {
                message = mTicket.ticket.holder_name + " יצא/ה בהצלחה מהאירוע."
            }
            val builder = AlertDialog.Builder(this@ShowActivity)
            builder.setMessage(message)
                    .setTitle("אישור")
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        val intent = Intent(this@ShowActivity, MainActivity::class.java)
                        startActivity(intent)
                    }
            val dialog = builder.create()
            dialog.show()
        }

        override fun onFailure(throwable: Throwable) {
            runOnUiThread { progressBar_ShowActivity.visibility = View.GONE }
            throwable.printStackTrace()
            AppUtils.playMusic(this@ShowActivity, AppConsts.ERROR_MUSIC)
            AlertDialog.Builder(this@ShowActivity).setTitle("פעולה נכשלה")
                    .setMessage("")
                    .setPositiveButton(getString(R.string.ok), null)
                    .setNegativeButton("", null)
                    .setIcon(android.R.drawable.ic_dialog_alert)
                    .show()
        }
    }


    private enum class Action {
        ENTER,
        EXIT
    }

    fun exit() {
        val hasInternetConnection = AppUtils.isConnected(this)
        if (!hasInternetConnection) {
            AlertDialog.Builder(this).setTitle(getString(R.string.no_network_dialog_title))
                    .setMessage(getString(R.string.no_network_dialog_message))
                    .setPositiveButton(getString(R.string.ok), null)
                    .setNegativeButton(null, null)
                    .setIcon(android.R.drawable.ic_dialog_alert)
                    .show()
            return
        }

        progressBar_ShowActivity.visibility = View.VISIBLE

        val barcode = mTicket.ticket.barcode
        Log.d(AppConsts.TAG, "user barcode to exit: $barcode")

        NetworkApi.gateExit(this, mGateCode, barcode, gateCallback)
    }

    fun entrance() {
        val hasInternetConnection = AppUtils.isConnected(this)
        if (!hasInternetConnection) {
            AlertDialog.Builder(this).setTitle(getString(R.string.no_network_dialog_title))
                    .setMessage(getString(R.string.no_network_dialog_message))
                    .setPositiveButton(getString(R.string.ok), null)
                    .setNegativeButton(null, null)
                    .setIcon(android.R.drawable.ic_dialog_alert)
                    .show()
            return
        }
        when (mTicket.gate_status) {
            TicketNew.State.EARLY_ENTRANCE -> handleGroupTypes()
            TicketNew.State.MIDBURN -> {
                NetworkApi.gateEnter(this, mGateCode, mTicket.ticket.barcode, gateCallback)
            }
        }
    }

    private fun handleGroupTypes() {
        val groups = mTicket.ticket.groups

        //check if group type is production. if so, select it immediately
        if (mTicket.ticket.production_early_arrival) {
            NetworkApi.gateEnter(this, mGateCode, mTicket.ticket.barcode, gateCallback, PRODUCTION_EARLY_ARRIVAL_WORKAROUND)
        }

        // no groups alert
        if (groups.isEmpty()) {
            AlertDialog.Builder(this).setTitle("שגיאה")
                    .setMessage(getString(R.string.no_early_arrival_message))
                    .setPositiveButton(getString(R.string.ok), null)
                    .setNegativeButton(null, null)
                    .setIcon(android.R.drawable.ic_dialog_alert)
                    .show()
            return
        }

        // show group selection dialog
        val groupsArrayListSize = groups.size
        val groupsArray = arrayOfNulls<CharSequence>(groupsArrayListSize)
        for (i in 0 until groupsArrayListSize) {
            val group = groups[i]
            groupsArray[i] = getGroupType(group) + ": " + group.name
        }
        val builder = AlertDialog.Builder(this)
        builder.setTitle("בחר קבוצה")
        builder.setItems(groupsArray) { _, which ->
            val selectedGroup = groups[which]
            Log.d(AppConsts.TAG, selectedGroup.name + " was clicked. id: " + selectedGroup.id)
            progressBar_ShowActivity.visibility = View.VISIBLE
            mSelectedGroup = selectedGroup
            NetworkApi.gateEnter(this, mGateCode, mTicket.ticket.barcode, gateCallback, selectedGroup.id)
        }
        builder.show()
    }

    private fun getGroupType(group: com.midburn.gate.midburngate.network.Group): String {
        if (group.type == AppConsts.GROUP_TYPE_ART) {
            return "מיצב"
        }

        if (group.type == AppConsts.GROUP_TYPE_CAMP) {
            return "מחנה"
        }

        return if (group.type == AppConsts.GROUP_TYPE_PRODUCTION) {
            "הפקה"
        } else ""
    }

    fun cancel() {
        val intent = Intent(this@ShowActivity, MainActivity::class.java)
        startActivity(intent)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_show)
        exitButton.setOnClickListener { exit() }
        entranceButton.setOnClickListener { entrance() }
        cancelButton.setOnClickListener { cancel() }
        supportActionBar?.title = getString(R.string.ticket_details)

        mGateCode = intent.getStringExtra("event_id") ?: AppUtils.getEventId(this) ?: throw NullPointerException("event_id must be non null")

        mTicket = intent.getSerializableExtra("ticketDetails") as TicketNew
        orderNumberTextView_ShowActivity.text = (intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.order_id.toString()
        ticketNumberTextView_ShowActivity.text = (intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.ticket_number.toString()
        ticketOwnerTextView_ShowActivity.text = (intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.holder_name
        ticketTypeTextView_ShowActivity.text = (intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.type
        ticketOwnerIdTextView_ShowActivity.text = (intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.israeli_id

        if (mTicket.ticket.production_early_arrival) {
            findViewById<View>(R.id.earlyArrivalProductionTV).visibility = View.VISIBLE
        }

        //decide which button to show (entrance/exit)
        when {
            (intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.inside_event == 0 -> //the user is outside the event
                mAction = Action.ENTER
            (intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.inside_event == 1 -> //the user is inside the event
                mAction = Action.EXIT
            else -> Log.e(AppConsts.TAG, "unknown isInsideEvent state. isInsideEvent: " + (intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.inside_event)
        }
        toggleButtonsState()


        //decide if disabled layout should be displayed
        if ((intent.getSerializableExtra("ticketDetails") as TicketNew).ticket.disabled_parking == 1) {
            //show disabled parking
            disabledLayout_ShowActivity.visibility = View.VISIBLE
        } else {
            disabledLayout_ShowActivity.visibility = View.GONE
        }

    }

    private fun toggleButtonsState() {
        if (mAction == Action.EXIT) {
            entranceButton.visibility = View.GONE
            exitButton.visibility = View.VISIBLE
        } else {
            exitButton.visibility = View.GONE
            entranceButton.visibility = View.VISIBLE
        }
    }

    override fun onBackPressed() {
        val upIntent = Intent(this, MainActivity::class.java)
        NavUtils.navigateUpTo(this, upIntent)
    }

    companion object {

        private const val PRODUCTION_EARLY_ARRIVAL_WORKAROUND = -1
    }
}
